csharp/src/Ice/Proxy.cs

```diff
@@ -322,86 +325,166 @@ public static class Proxy
-                        catch (OperationCanceledException)
+                        catch (TransportException ex) when (request.IsIdempotent || !progressWrapper.IsSent)
                         {
-                            throw; // Don't retry cancelled operations
+                            // Retry transport exceptions if the request is idempotent or was not send
+                            lastException = ex;
+                            retryable = Retryable.AfterDelay;
                         }
-                        catch (Exception ex)
+
+                        if (lastException == null)
                         {
-                            // Clear the proxy's cached request handler if connection caching is enabled
-                            if (reference.IsConnectionCached && handler != null)
+                            Debug.Assert(response != null && response.ResultType == ResultType.Failure);
+                            observer?.RemoteException();
+                            if (response.ReadSystemException(proxy.Communicator) is Exception systemException &&
+                                systemException is ObjectNotExistException one)
                             {
-                                proxy.IceReference.ClearRequestHandler(handler);
+                                // 1.1 System exceptions
+                                lastException = systemException;
+                                if (reference.RouterInfo != null && one.Operation == "ice_add_proxy")
+                                {
+                                    // If we have a router, an ObjectNotExistException with an operation name
+                                    // "ice_add_proxy" indicates to the client that the router isn't aware of the proxy
+                                    // (for example, because it was evicted by the router). In this case, we must
+                                    // *always* retry, so that the missing proxy is added to the router.
+                                    reference.RouterInfo.ClearCache(reference);
+                                    retryable = Retryable.AfterDelay;
+                                }
+                                else if (reference.IsIndirect)
+                                {
+                                    if (reference.IsWellKnown)
+                                    {
+                                        reference.LocatorInfo?.ClearCache(reference);
+                                    }
+                                    retryable = Retryable.AfterDelay;
+                                }
                             }
+                            else if (response.BinaryContext.TryGetValue((int)BinaryContext.RetryPolicy,
+                                                                        out ReadOnlyMemory<byte> value))
+                            {
+                                retryable = (Retryable)value.Span[0];
```

Comment:
Why not unmarshal it cleanly with an InputStreamReader?

Reply:
Fine I will add an unmarshal constructor to RetryPolicy also, so we can keep the properties read-only